File: baekjoon/binary-search/java/BOJ2632.java
import java.util.*;
import java.io.*;
public class BOJ2632 {
    static int T, N, M;
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        T = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        int[] A = new int[N];
        int[] B = new int[M];

        for(int i=0;i<N;i++){
            A[i] = Integer.parseInt(br.readLine());
        }
        for(int i=0;i<M;i++){
            B[i] = Integer.parseInt(br.readLine());
        }

        int[] stackSumA = getStackSum(A);
        int[] stackSumB = getStackSum(B);

        Map<Integer, Integer> ACount = getCountAsMap(stackSumA);
        Map<Integer, Integer> BCount = getCountAsMap(stackSumB);
        binarySearch(ACount, BCount);
    }
    public static int[] getStackSum(int[] arr){
        ArrayList<Integer> res = new ArrayList<>();
        int start = 0;
        int end = arr.length - 1; 
        while(true){
            int i = start;
            int sum = 0;
            while(i != end){
                sum += arr[i];
                if(sum > T) break;
                res.add(sum);

                i = (i + arr.length + 1) % arr.length;
            }

            start = (start + arr.length + 1) % arr.length;
            end = (start + arr.length - 1) % arr.length;
            if(start == 0) break;
        }
        int sum = 0;
        for(int i=0;i<arr.length;i++){
            sum += arr[i];
        }
        if(sum <= T){
            res.add(sum);
        }
        int[] returnValue = new int[res.size()];
        for(int i=0;i<res.size();i++){
            returnValue[i] = res.get(i);
        }
        
        return returnValue;
    }
    public static void binarySearch(Map<Integer, Integer> ACount, Map<Integer, Integer> BCount){
        ArrayList<Integer> arr1 = new ArrayList<>(ACount.keySet());
        ArrayList<Integer> arr2 = new ArrayList<>(BCount.keySet());

        Collections.sort(arr1);
        Collections.sort(arr2);

        int left = 0, right = arr2.size() - 1;
        int res = ACount.getOrDefault(T, 0) + BCount.getOrDefault(T, 0);

        while(left < arr1.size() && right >= 0){
            int cur = arr1.get(left) + arr2.get(right);

            if(cur > T){
                right--;
            } else if(cur < T){
                left++;
            } else {
                res += (ACount.get(arr1.get(left)) * BCount.get(arr2.get(right)));
                left++;
                right--;
            }
        }
        System.out.println(res);
    }
    public static Map<Integer, Integer> getCountAsMap(int[] arr){
        Map<Integer, Integer> count = new HashMap<>();

        for(int num : arr){
            count.put(num, count.getOrDefault(num, 0) + 1);
        }
        return count;
    }
}

File: 0_1.baekjoon/binary-search/python/BOJ2632.py
from sys import stdin
from collections import defaultdict
readline = stdin.readline

T = int(readline())
N, M = map(int, readline().split())

A = [int(readline()) for _ in range(N)]
B = [int(readline()) for _ in range(M)]



def solve():
    stack_sum_A = list(sorted(get_stack_sum(A)))
    stack_sum_B = list(sorted(get_stack_sum(B)))

    A_count = get_count_as_dict(stack_sum_A)
    B_count = get_count_as_dict(stack_sum_B)

    res = binary_search(A_count, B_count)
    print(res)

def get_count_as_dict(arr):
    count = defaultdict(int)
    for a in arr:
        if a not in count.keys():
            count[a] = 0
        count[a] += 1
    return count

def binary_search(dict1, dict2):
    arr1 = list(sorted(list(dict1.keys())))
    arr2 = list(sorted(list(dict2.keys())))

    left, right = 0, len(arr2) - 1
    res = dict1[T] + dict2[T]
    while left < len(arr1) and right >= 0:
        cur = arr1[left] + arr2[right]

        if cur > T:
            right -= 1
        elif cur < T:
            left += 1
        else:
            res += (dict1[arr1[left]] * dict2[arr2[right]])
            left += 1
            right -= 1
    return res

# 누적합
# Circular
def get_stack_sum(arr):
    res = []
    start = 0
    end = len(arr) - 1
    # 순환 배열 누적합
    while True:
        _sum = 0
        i = start
        while i != end:
            _sum += arr[i]
            if _sum > T: break
            res.append(_sum)

            i = (i + len(arr) + 1) % len(arr)

        start = (start + len(arr) + 1) % len(arr)
        end = (start + len(arr) - 1) % len(arr)
        # 시작점이 다시 처음으로 돌아오면 종료
        if start == 0: break
    
    _sum = sum(arr)
    if _sum <= T:
        res.append(_sum)
    return res

solve()